### ISSUE TITLE
Add extra include to library root module

### ISSUE THREAD

**Issue by Pitometsu** [ocaml/dune#2156]
Suppose, I have such librarary `m` in project`p`:
```
m
├── dune
├── a.ml
└── b.ml
```
`dune`:
```lisp
(library
  (name m)
  (public_name p.m))
```
AFAIK, it would be the  same as have  
`m.ml`:
```
module A = М__А
module B = M__B
```
***
But what if I want to have there also
`m.ml`:
```
(* ... *)
module type A = module type of М__А
module type B = module type of M__B
```
I know, I have 2 ways: let `m.ml` be generated by `dune` and create *nested* module (e.g. name it `s.ml`) with signature names.
Or do it as desired right in the `M` namespace (like `include S` private module). But then we would be forced to declare the whole `M` explicitly together with including of all the public sub-modules, loose `dune`'s handy help with it. A question is: can it be avoided, is there any known workarounds, and best workflow practices how to add additional mix-in to top library level in `dune`?

**Comment by Pitometsu** [comment]
I mean, is there a way to avoid list each submodule in `m.ml` explicitly by hand if I want just to add something there?

**Comment by diml** [comment]
It's not possible no. I understand that this can be a bit annoying to have to write this list manually, but on the other hand the current scheme has the benefit of being simple: either you export all the modules and only the modules, either you have a completely explicit interface. If we allow to have a mix of both, it becomes hard to follow what's happening.

What I would recommend is to add a test to make sure the explicit list is up to date, and use the promotion feature of dune to automatically update it if it is not. This way, you get the best of both world: you don't have to maintain the list explicitly but things are still clear and explicit. The feature is described [here](https://dune.readthedocs.io/en/latest/dune-files.html#diffing-and-promotion) in the manual, but in a nutshell you'd write something like this:

```scheme
(rule
 (targets m.ml.corrected)
 (deps (:mlfiles (glob_files *.ml)))
 (action (with-stdout-to m.ml.corrected (run gen-m/gen.exe %{mlfiles})))

(alias
 (name runtest)
 (action (diff m.ml m.ml.corrected)))
```

Where `gen-m/gen.exe` is a program that reads `m.ml`, update the list of module aliases or types according to the list of files it gets as argument, and output an updated version of `m.ml`. When you add or remove a .ml file, `dune runtest` would show you that `m.ml` and `m.ml.corrected` don't match, and you'd be able to run `dune promote` to automatically replace `m.ml` in your source tree by the generated `m.ml.corrected`.

**Comment by Pitometsu** [comment]
Can I do something similar in build dir on build phase?
E.g. rename generated by dune `m.ml` to `m__m_.ml`, rename `s.ml` to `m.mr`, where `s.ml` contain `include M_`?

**Comment by Pitometsu** [comment]
Something like (this does not work yet):
`dune`:
```lisp
(library
  (name M)
  (private_modules m_ s)
;; ...

(rule
 (targets m_.ml)
 (action (copy# m__.ml-gen m_.ml)))

(rule
 (targets m.ml)
 (action (copy# s.ml m.ml)))
```
`s.ml`:
```ocaml
include M_
(* ... *)
```

**Comment by Pitometsu** [comment]
But this one looks like working:
```lisp
(rule
 (targets m.ml)
 (action (with-stdout-to m.ml
                         (progn
                           (echo "include M__\n\n")
                           (cat s.ml)))))
```
Any potential drawbacks here? @diml

**Comment by Pitometsu** [comment]
Or even simpler, with no rules, just by:
`m.ml`:
```ocaml
include M__

include S
```

Is it undocumented behaviour? Could it be a feature?

**Comment by Pitometsu** [comment]
Feature request:
- or freeze `<Library_name>__` as expected to enhance in documentation
- or add to library stanza some bool flag like `enhanceable` to `echo "include M__\n"` at the head of `<library_name>.ml` on building phase.

**Comment by Pitometsu** [comment]
Benefits:
- cleaner module system DSL without per-library workarounds for signatures and types
- much more obvious `modules_without_implementation`

**Comment by Pitometsu** [comment]
Another drawback: that additional declarations in libraries toplevel can't be used by library itself. :disappointed:

**Comment by Pitometsu** [comment]
Well, for this case `S` could be in sub-library (name it lib) and open like `-open S` in dune's `(flags`.

**Comment by diml** [comment]
It is not recommended to rely on the  `Foo__` and `Foo__bar` names. We are expecting that one day OCaml will have proper support for namespaces. When this happens, Dune will use namespaces rather than generate these modules. In general, things that are not documented are not considered stable.

Maybe we could add a form `%{module-names:<lib>}` that would expand to the list of modules of `<lib>`. Then it would be even easier to generate your module.

**Comment by Pitometsu** [comment]
General goal is to add types/signatures (include submodules into top library module rather then just add it).
